src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/FSharp/FSharpValueOptionConverter.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json.Serialization.Metadata;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    // Converter for F# struct optional values: https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-fsharpvalueoption-1.html
+    internal sealed class FSharpValueOptionConverter<TValueOption, TElement> : JsonResumableConverter<TValueOption>
+        where TValueOption : struct, IEquatable<TValueOption>
+    {
+        // While technically not implementing IEnumerable, F# optionals are effectively generic collections of at most one element.
+        internal override ConverterStrategy ConverterStrategy => ConverterStrategy.Enumerable;
+        internal override Type? ElementType => typeof(TElement);
+
+        public override bool HandleNull => true;
+
+        private readonly FSharpCoreReflectionProxy.StructGetter<TValueOption, TElement> _optionValueGetter;
+        private readonly Func<TElement?, TValueOption> _optionConstructor;
+
+        [RequiresUnreferencedCode(FSharpCoreReflectionProxy.FSharpCoreUnreferencedCodeMessage)]
+        public FSharpValueOptionConverter(JsonConverter<TElement> elementConverter)
+        {
+            _optionValueGetter = FSharpCoreReflectionProxy.Instance.CreateFSharpValueOptionValueGetter<TValueOption, TElement>();
+            _optionConstructor = FSharpCoreReflectionProxy.Instance.CreateFSharpValueOptionSomeConstructor<TValueOption, TElement>();
+            // If the element converter is value, this converter will also be writing values
+            // Set a flag to signal this fact to the converter infrastructure.
+            CanWriteJsonValues = elementConverter.ConverterStrategy == ConverterStrategy.Value;
```

Comment:
Instead of adding a new `CanWriteJsonValues` property, could we set `ConverterStrategy` above on L15 according to the elementConverter value (when applicable)?

Reply:
That was my initial approach, however it turned out that if the elementConverter is `ConverterStrategy.Value`, reflecting that on the option converter resulted in a few breaks IIRC `TryWrite` was not being called resulting in inappropriate null value handling as well as new stackframes not being allocated.

Reply:
I had a thought about this, and eventually made a few amendements in https://github.com/dotnet/runtime/pull/55108/commits/17a76a08d227049234b7f55046d56ae367951ae3 following your feedback. Ultimately it results in a different set of infrastructural changes, but I feel they're less intrusive.